### ISSUE TITLE
Spack PR testing

### ISSUE THREAD

**Issue by tgamblin** [spack/spack#7117]
In addition to the release testing described in #7116, Spack should do build tests on pull requests made to `develop`.

- [ ] Create a binary package repository with builds for the latest `develop`.

- [ ] Implement PR testing for develop in a similar way to releases -- treat `develop` as an "ongoing" release, and it should have a default set of test builds defined in a `release.yaml` (or maybe some other name) a the root of the repository.

- [ ] Develop a suitable purge policy for `develop`.  We can't keep around all builds as binary packages, but we can keep around some. 

- [ ] Implement the ability for authorized users to comment on PRs and trigger builds of specified dependents of the packages in the PR.  A user with Spack write access should be able to comment on a PR with syntax like `build: <spec>` and trigger the build of the specified spec.

**Comment by mathstuf** [comment]
Cc: @opadron

**Comment by goxberry** [comment]
Two  comments based on issues I've run into while fixing up packages:

- It would be worth running `ldd` or `otool -L` on the libraries built by spack package and making sure that all `link` dependencies are actually linked to the libraries that were built. (This diagnostic might also apply to executables?) Sometimes, packages rely on spack modules setting environment variables so that Autotools or CMake "does the right thing", detects the correct version of a dependency, and links it. But I've run into cases where the build system instead detects and pulls in a system version of a dependency instead of the spack version, and it's a pain in the ass to debug. This mode of breakage definitely happened when debugging #7039, and possibly caused the bug in #7182.

- By the same token, it would also be worth running a post-install test on packages where possible and practical. Spack can successfully install a package, and then linking it and running an executable will fail because shared libraries can't resolve symbols that were left out. This issue was another surprise with `mfem` in #7165, and I wouldn't have discovered it if I weren't trying to add the `glvis` package in #7166. IIRC, there's an unstable API for these sorts of tests, so coming up with a stable API would probably encourage more contributors to add build time tests. Running `make test` doesn't necessarily test a correct install -- sometimes it tests the pre-install build, so care must be taken in implementing these tests, but usually, other code (e.g., libraries built with examples, or build-time tests) can be reused for this task.

**Comment by mathstuf** [comment]
> `ldd` or `otool -L`

That could work, but some places use system dependencies. I suppose that this can be an option for those generating binaries meant to be redistributed though. There may need to be a whitelist of OK-to-use system paths (for platform compilers or libraries).

> But I've run into cases where the build system instead detects and pulls in a system version of a dependency instead of the spack version

This usually indicates a problem with the build system of the package being built (either being too flexible and finding mismatches or just looking in hard-coded locations).

> it would also be worth running a post-install test on packages where possible and practical

Some projects do have test suites can be run against their install trees (VTK for example). Others rely on files only available in the build tree or on linking backing libraries for access to private symbols (IMO, bad design, but it's not something we can ignore). But if we're making sanity tests for packages, maybe it's best to make a package which depends on the built package and add it to the DAG if requested? For example, the `py-numpy-test` package would depend on `py-numpy` and make sure that importing it works as well as compiling a small library against the headers. That also tests that Spack makes the package available properly rather than just testing that the install is useful.

**Comment by goxberry** [comment]
> This usually indicates a problem with the build system of the package being built (either being too flexible and finding mismatches or just looking in hard-coded locations).

Agreed, and build systems typically cope with badly designed build systems.

>Some projects do have test suites can be run against their install trees (VTK for example). Others rely on files only available in the build tree or on linking backing libraries for access to private symbols (IMO, bad design, but it's not something we can ignore).

See my first comment in this reply.

> But if we're making sanity tests for packages, maybe it's best to make a package which depends on the built package and add it to the DAG if requested? For example, the `py-numpy-test` package would depend on `py-numpy` and make sure that importing it works as well as compiling a small library against the headers. That also tests that Spack makes the package available properly rather than just testing that the install is useful.

Right now, the only way to test e.g., `mfem` for installation bugs is to write a package that depends on it, e.g., `glvis`. It's unclear to me what a separate package buys you, because then you're also relying on the concretizer to pick the right `py-numpy` to go with `py-numpy-test`. It seems like all of the information necessary to test an installation could be contained within the package itself, and setup/run/teardown could be done in a separate `test` phase of a build (which, as we've both noted, is already done in some cases). Homebrew already does this.

I'm wary of adding "test-only" packages because it seems like a lot of noise in the directory tree, and separating test info into its own package raises more synchronization issues than putting it in the same file. And I'm wary of the concretization issue, because I occasionally run into edge cases with it, or it sometimes installs a bunch of duplicates of packages, even if I attempt to force dependencies.

**Comment by mathstuf** [comment]
> Agreed, and build systems typically cope with badly designed build systems.

I think Spack can be used to help push upstreams to fixing their build systems rather than just coping with them. Future you will be happier :) .

> I'm wary of adding "test-only" packages

Maybe they could be "hidden" in some way (maybe `numpy/test/package.py`?). As for synchronization issues, building it and making it depend on `^numpy/<hash>` that was just built should work.

**Comment by tgamblin** [comment]
This is resolved by a lot of work by @zackgalbreath, @scottwittenburg, and @opadron over in https://github.com/spack/spack-infrastructure
